index.js

```diff
@@ -28,230 +25,266 @@ const IMPORT_PATH_DEFAULT = {
-  ) {
-    try {
-      return leadingComments[0].value
-        .split('webpackChunkName:')[1]
-        .replace(/["']/g, '')
-        .trim()
-    }
-    catch (e) {
-      return null
+function idOption(p, t) {
+  const id = getComponentId(p)
+  return t.objectProperty(t.identifier('id'), t.stringLiteral(id))
+}
+
+function getWebpackComments({ node: { leadingComments = [] } }) {
+  return leadingComments.reduce((comments, { value }) => {
+    const isMagicComment = validMagicStrings
+      .filter(str => new RegExp(`${str}\\w*:`).test(value))
+      .length === 1
+    if (isMagicComment) {
+      comments.push(value)
     }
-  }
-  return null
+    return comments
+  }, [])
 }
 
-function idOption(t, importArgNode) {
-  const id = getComponentId(t, importArgNode)
-  return t.objectProperty(t.identifier('id'), t.stringLiteral(id))
+function trimChunkNameBaseDir(baseDir) {
+  return baseDir.replace(/^[./]+|(\.js$)/g, '')
 }
 
-function fileOption(t, p) {
-  return t.objectProperty(
-    t.identifier('file'),
-    t.stringLiteral(
-      path.relative(__dirname, p.hub.file.opts.filename || '') || ''
-    )
-  )
+function createWebpackChunkName({ node: { expressions, quasis, value } }) {
+  if (!quasis) {
+    return trimChunkNameBaseDir(value)
+  }
+
+  const [{ value: { cooked: baseDir } }] = quasis
+  const chunkName = baseDir + (expressions.length > 0 ? '[request]' : '')
+
+  return trimChunkNameBaseDir(chunkName)
 }
 
-function loadOption(t, loadTemplate, p, importArgNode) {
-  const argPath = getImportArgPath(p)
-  const generatedChunkName = getMagicCommentChunkName(importArgNode)
-  const otherValidMagicComments = getMagicWebpackComments(importArgNode)
-  const existingChunkName = t.existingChunkName
-  const chunkName = existingChunkName || generatedChunkName
-
-  delete argPath.node.leadingComments
-  argPath.addComment('leading', ` webpackChunkName: '${chunkName}' `)
-  otherValidMagicComments.forEach(validLeadingComment =>
-    argPath.addComment('leading', validLeadingComment.value)
-  )
+function loadOption(p, t, template, chunkName) {
+  const magicComments = [
+    ` webpackChunkName: '${chunkName || createWebpackChunkName(p)}' `,
+    ...getWebpackComments(p)
+  ]
+  delete p.node.leadingComments
+  magicComments.forEach(comment => p.addComment('leading', comment))
+  const load = template({ IMPORT: p.parent })
+  return t.objectProperty(t.identifier('load'), load.expression)
+}
 
-  const load = loadTemplate({
-    IMPORT: argPath.parent
-  }).expression
+function getImport(p, { source, nameHint }) {
+  return addDefault(p, source, { nameHint })
+}
 
-  return t.objectProperty(t.identifier('load'), load)
+function getWebpackChunkName({ node: { leadingComments = [] } }) {
+  const chunkNameComment = leadingComments.find(({ value }) => value.includes('webpackChunkName:'))
+  if (chunkNameComment) {
+    return chunkNameComment.value
+      .split('webpackChunkName:')[1]
+      .replace(/["'\s]/g, '')
+  }
+  return null
```

Comment:
Previously named `existingMagicCommentChunkName`. This rewrite fixes an issue with a wrong result expected in [this test case](https://github.com/faceyspacey/babel-plugin-universal-import/blob/master/__tests__/__snapshots__/index.test.js.snap#L186), renamed to `import with custom webpackChunkName + strips out unknown things`, where the chunk name is `Foo` (filename) instead of `Bar` (magic comment value).